### DISCUSSION TITLE
Interface ID Calculation

### DISCUSSION THREAD

**Discussion by ghost** [OpenZeppelin/cairo-contracts#923] [Q&A] [answered]
Hi everyone. I'm wondering that how can the interfaces id being calculated (functions that has a State parameter), like in this code: https://github.com/OpenZeppelin/cairo-contracts/blob/main/src/token/erc721/interface.cairo. I'm following this docs: https://github.com/starknet-io/SNIPs/blob/main/SNIPS/snip-5.md#how-interfaces-are-identified, but the example interface used doesn't have a State parameters. I tried adding `@TState` into the param fields but the result returned didn't match the one calculated in the erc721 interface file.

**Comment by ericnordelo** [comment]
Interface IDs are computed from the non-generic version of the interface, (removing self) which is the one matching the external ABI of the contract. You should remove the generic TContractState from the interface trait before computing the ID, and this is a requirement for [src5-rs](https://github.com/ericnordelo/src5-rs) as well.

**Reply by ericnordelo** [reply]
Notice that the real external interface of the contract doesn't have any generic param, so `self: ContractState` is a language feature for handling visibility, but is not part of the ABI. That's why is not included.

**Reply by ericnordelo** [reply, marked as answer]
https://github.com/ericnordelo/src5-rs This is a tool for computing the interface id in case it can help.
